File: code/java/lists/src/main/java/org/swufe/datastructure/LinkedStack.java
package org.swufe.datastructure;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * A stack based on singly linked lists.
 */
public class LinkedStack<Item> implements Iterable<Item> {
    private static class Node<Item> {
        Item item;
        Node<Item> next;

        Node(Item item) {
            this.item = item;
            next = null;
        }
    }

    private Node<Item> head;
    private int size;

    public LinkedStack() {
        head = null;
        size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    // push() = add_first()
    public void push(Item item) {
        Node<Item> node = new Node<>(item);
        node.next = head;
        head = node;
        size += 1;
    }

    // pop() = remove_first()
    public Item pop() {
        if (isEmpty()) {
            throw new NoSuchElementException("Pop from an empty stack");
        }
        Item answer = head.item;
        head = head.next;
        size -= 1;
        return answer;
    }

    public Item peek() {
        if (isEmpty()) {
            throw new NoSuchElementException("Peek from an empty stack");
        }
        return head.item;
    }

    @Override
    public Iterator<Item> iterator() {
        return new LinkedIterator();
    }

    private class LinkedIterator implements Iterator<Item> {
        private Node<Item> node = head;

        @Override
        public boolean hasNext() {
            return node != null;
        }

        @Override
        public Item next() {
            Item answer = node.item;
            node = node.next;
            return answer;
        }
    }
}

File: code/python/lists/linked_stack.py
from no_element import NoElement


class Node:
    def __init__(self, item, nxt=None):
        self.item = item
        self.next = nxt


class LinkedStack:
    """A stack based on singly linked list."""

    def __init__(self):
        self._head = None
        self._size = 0

    def size(self):
        return self._size

    def is_empty(self):
        return self._size == 0

    # push() = add_first()
    def push(self, item):
        node = Node(item)
        node.next = self._head
        self._head = node
        self._size += 1

    # pop() = remove_first()
    def pop(self):
        if self.is_empty():
            raise NoElement('Pop from an empty stack')
        answer = self._head.item
        self._head = self._head.next
        self._size -= 1
        return answer

    def peek(self):
        if self.is_empty():
            raise NoElement('Peek from an empty stack')
        return self._head.item

    def __iter__(self):
        return LinkedIterator(self._head)


class LinkedIterator:
    def __init__(self, head: Node):
        self.walk = head

    def __next__(self):
        if self.walk is None:
            raise StopIteration
        answer = self.walk.item
        self.walk = self.walk.next
        return answer
